File: tools/commonist/src/scala/commonist/task/UploadFilesTask.scala
package commonist.task

import java.io.File

import javax.swing.JOptionPane

import scutil.Log._
import scutil.TextFile
import scutil.SwingUtil._
import scutil.ext.FileExt._

import commonist.Task
import commonist.Parser
import commonist.Constants
import commonist.api._
import commonist.data._
import commonist.task.upload._
import commonist.ui.MainWindow
import commonist.ui.ImageListUI
import commonist.ui.StatusUI
import commonist.ui.later._
import commonist.util.Loader
import commonist.util.Messages
import commonist.util.TextUtil2
import commonist.util.JavaCollectionUtil

/** upload files selected in the ImageListUI */
class UploadFilesTask(
	settingsDir:File,
	loader:Loader, 
	commonData:CommonData, 
	imageListData:ImageListData, 
	mainWindow:MainWindow,
	imageListUI:ImageListUI,
	statusUI:StatusUI 
) extends Task {
	private val statusUILater		= new StatusUILater(statusUI)
	private val imageListUILater	= new ImageListUILater(imageListUI)
			
	private val	wiki			= commonData.wiki
	private val	api				= new API(wiki.api)
	private val wikiName		= wiki.toString
	private val uploadTemplates	= new UploadTemplates(loader, wiki)
	
	override protected def execute() {
		if (!imageListData.hasSelected) {
			INFO("nothing to upload")
			statusUILater.halt("status.upload.empty")
			return
		}
		
		try {
			if (!login())	return
			val common	= Common(
				commonData.description.trim,
				commonData.date.trim,
				commonData.source.trim,
				commonData.author.trim,
				commonData.license.template,
				commonData.license.description,
				Parser parseCategories commonData.categories
			)
			val	uploads	= upload(common)
			if (Constants.ENABLE_GALLERY) {
				gallery(common, uploads)
			}
			INFO("upload finished")
		}
		catch { 
			case e:AbortedException =>
				ERROR("upload task aborted")
				statusUILater.halt("status.upload.aborted")
			case e:Exception =>
				ERROR("upload task error", e)
				// TODO hack
				// status.upload.error=Hochladen von {0} fehlgeschlagen ({1})
				statusUILater.halt("status.upload.error", "", e.getMessage)
		}
	}
	
	/*
	statusUILater.halt("status.login.successful", wikiName)
	statusUILater.halt("status.login.aborted")
	statusUILater.halt("status.login.error", wikiName, e.getMessage)
	statusUILater.halt("status.login.wrongpw", wikiName)
	*/
	private def login():Boolean = {
		INFO("logging in")
		statusUILater.indeterminate("status.login.started", wikiName)
		check()
			
		val loginResult	= api.login(commonData.user.trim, commonData.password)
		loginResult match {
			case LoginSuccess(userName)	=>
				INFO("login successful: " + userName)
				statusUILater.halt("status.login.successful", wikiName)
				true
			case LoginFailure(code)	=>
				INFO("login failed: " + code)
				// TODO more detail
				statusUILater.halt("status.login.wrongpw", wikiName)
				false
			case LoginError(code)	=>
				INFO("login error: " + code)
				statusUILater.halt("status.login.error", wikiName, code)
				false
		}
	}
	
	/*
	statusUILater.halt("status.upload.aborted")
	statusUILater.halt("status.upload.error", path, e.getMessage)
	*/
	private def upload(common:Common):List[Upload] = {
		INFO("uploading files")
		
		// TODO normalizeTitle(FileName.fix( is stupid
		val	selected	= imageListData.selected
		def titleAt(index:Int):String	=
					 if (index < 0)					null
				else if (index >= selected.size)	null
				else Namespace.file(Filename.normalizeTitle(Filename.fix(selected(index).name)))
				
		selected.zipWithIndex map { case Pair(imageData, index) => 
			check()
			
			val	file		= imageData.file
			val fileLength	= file.length
			val fileName	= file.getName
			val filePath	= file.getPath
			val	name		= Filename.normalizeTitle(Filename.fix(imageData.name))
			val title		= Namespace.file(name)
			val previous	= titleAt(index-1)
			val next		= titleAt(index+1)
			val coords		= imageData.coordinates
			val coordParts	= Parser parseCoordinates imageData.coordinates
			val latitude	= coordParts map { _._1 } getOrElse null
			val longitude	= coordParts map { _._2 } getOrElse null
			val categories	= Parser parseCategories imageData.categories
			
			val upload	= Upload(
				name,
				title,
				null,
				previous,
				next,
				imageData.description.trim,
				imageData.date.trim,
				imageData.permission.trim,
				categories,
				coords.trim,
				latitude,
				longitude
			)
			
			statusUILater.halt("status.upload.started", fileName)
			val callback	= new MyUploadCallback(mainWindow, statusUILater, fileLength, fileName, name)
			val text		= uploadTemplates.imageDescription(common, upload)
			val	watch		= true
			
			val uploaded	= api.upload(name, "", text, watch, file, callback)
			uploaded match {
				case UploadSuccess(fileName, pageTitle)	=> 
					INFO("upload successful: " + fileName + " to " + pageTitle)
					statusUILater.halt("status.upload.successful", fileName, pageTitle)
					imageListUILater.uploadFinished(file, true)
					upload.copy(name=fileName, title=pageTitle)
				case UploadAborted	=>
					// TODO more detail
					ERROR("upload aborted: " + fileName)
					statusUILater.halt("status.upload.error", fileName, "aborted")
					imageListUILater.uploadFinished(file, false)
					// TODO just remove it from the list?
					upload.copy(error="aborted")
				case UploadFailure(code)	=>
					// TODO more detail
					ERROR("upload failed: " + fileName + " because " + code)
					statusUILater.halt("status.upload.error", fileName, code)
					imageListUILater.uploadFinished(file, false)
					upload.copy(error=code)
				case UploadError(code)	=>
					ERROR("upload error: " + fileName + " because " + code)
					statusUILater.halt("status.upload.error", fileName, code)
					imageListUILater.uploadFinished(file, false)
					upload.copy(error=code)
			}
		}
	}
	
	/*
	statusUILater.halt("status.gallery.error", e.getMessage)
	statusUILater.halt("status.gallery.editConflict",	"[[" + title + "]]")
	*/
	def gallery(common:Common, uploads:List[Upload]) = {
		INFO("changing gallery")
		check()
		
		val	title	= Namespace.user(commonData.user + "/gallery")
		val (sucesses, failures)	= uploads partition { _.error == null }
		
		val	batch	= Batch(
			JavaCollectionUtil.mkJavaList(uploads),
			JavaCollectionUtil.mkJavaList(sucesses),
			JavaCollectionUtil.mkJavaList(failures)
		)
		val summary	= uploadTemplates.gallerySummary(Constants.VERSION, failures.size)
		val text	= uploadTemplates.galleryDescription(common, batch)
		
		// backup gallery text
		val backup	= settingsDir / "gallery.txt"
		INFO("writing gallery to: " + backup)
		new TextFile(backup, "UTF-8").write(text)
		
		statusUILater.indeterminate("status.gallery.loading", "[[" + title + "]]")
		val editResult	= api.edit(title, summary, None, { oldText =>
			statusUILater.indeterminate("status.gallery.storing", "[[" + title + "]]")
			val newText	= text + "\n\n" + TextUtil2.trimLF(oldText)
			Some(newText)
		})
		editResult match {
			case EditSuccess(pageTitle)	=> 
				statusUILater.halt("status.gallery.updated",	"[[" + title + "]]")
			case EditAborted		=>
				// will not happen
			case EditFailure(code)	=> 
				// TODO more detail
				statusUILater.halt("status.gallery.error",		code + " in [[" + title + "]]")
			case EditError(code)	=> 
				statusUILater.halt("status.gallery.error",		code + " in [[" + title + "]]")
		}
	}
		
	/*
	statusUILater.halt("status.logout.error", wikiName, e.getMessage)
	if (!success) { statusUILater.halt("status.logout.failed", wikiName); return }
	*/
	private def logout() {
		INFO("logging out")
		statusUILater.indeterminate("status.logout.started", wiki)
		api.logout()
		statusUILater.halt("status.logout.successful", wikiName)
	}
	
	//==============================================================================
	
	/** asks the user when somwething about a file upload is unclear */
	private class MyUploadCallback(mainWindow:MainWindow, statusUILater:StatusUILater, fileLength:Long, fileName:String, name:String) extends UploadCallback {
		def progress(bytes:Long) {
			// System.err.println("written " + bytes + " of " + ofBytes)
			val percent	= (bytes * 100 / fileLength).toInt
			statusUILater.determinate("status.upload.progress", percent, 100, fileName, int2Integer(percent))
			// rate = (bytes - oldBytes) / (time  - oldTime)
		}
				
		/** when true, the file overwites another file */
		def ignoreFileexists():Boolean = callbackQuery(
				name,
				"query.upload.ignoreFileexists.title", 
				"query.upload.ignoreFileexists.message")	// false
		
		/** if true, previously uploaded and deleted files will be uploaded nevertheless */
		def ignoreFilewasdeleted():Boolean = callbackQuery(
				name,
				"query.upload.ignoreFilewasdeleted.title", 
				"query.upload.ignoreFilewasdeleted.message")	// true
		
		/** if true, already existing files will be uploaded nevertheless */
		def ignoreDuplicate():Boolean = callbackQuery(
				name,
				"query.upload.ignoreDuplicate.title", 
				"query.upload.ignoreDuplicate.message")
		
		/** if true, already existing files will be uploaded nevertheless */
		def ignoreDuplicateArchive():Boolean = callbackQuery(
				name,
				"query.upload.ignoreDuplicateArchive.title", 
				"query.upload.ignoreDuplicateArchive.message")
		
		/** ask the user a yes/no message */
		private def callbackQuery(name:String, titleKey:String, messageKey:String):Boolean ={
			try {
				edtWait {
						JOptionPane.YES_OPTION == JOptionPane.showConfirmDialog(
								mainWindow.window,
								Messages.message(messageKey, name),
								Messages.text(titleKey),
								JOptionPane.YES_NO_OPTION)
				}
			}
			catch { 
				case e:Exception =>
					ERROR("callback error", e)
					throw e
			}
		}
	}
}

File: tools/commonist-java/src/net/psammead/commonist/task/UploadFilesTask.java
package net.psammead.commonist.task;

import java.io.File;
import java.util.Iterator;
import java.util.List;

import javax.swing.JOptionPane;
import javax.swing.SwingUtilities;

import net.psammead.commonist.Task;
import net.psammead.commonist.data.CommonData;
import net.psammead.commonist.data.ImageData;
import net.psammead.commonist.data.ImageListData;
import net.psammead.commonist.task.edt.StatusUILater;
import net.psammead.commonist.text.Gallery;
import net.psammead.commonist.text.TemplateException;
import net.psammead.commonist.text.Templates;
import net.psammead.commonist.text.Upload;
import net.psammead.commonist.ui.MainWindow;
import net.psammead.commonist.ui.StatusUI;
import net.psammead.commonist.util.Messages;
import net.psammead.mwapi.Location;
import net.psammead.mwapi.MediaWiki;
import net.psammead.mwapi.NameSpace;
import net.psammead.mwapi.ui.Page;
import net.psammead.mwapi.ui.ProgressCallback;
import net.psammead.mwapi.ui.UnsupportedWikiException;
import net.psammead.mwapi.ui.UploadCallback;
import net.psammead.mwapi.ui.Uploaded;
import net.psammead.util.Logger;
import net.psammead.util.TextUtil;

/** upload files selected in the ImageListUI */
public class UploadFilesTask extends Task {
	private static final Logger log = new Logger(UploadFilesTask.class);
	
	private final MediaWiki	mw;
	private final MainWindow	mainWindow;
	private final CommonData 	commonData;
	private	final Gallery		gallery;
	
	private final StatusUILater	statusUILater;

	public UploadFilesTask(MediaWiki mw, Templates templates, MainWindow mainWindow, 
			CommonData commonData, ImageListData listData, 
			StatusUI statusUI) {
		this.mw				= mw;
		this.mainWindow		= mainWindow;
		this.commonData		= commonData;
		this.statusUILater	= new StatusUILater(statusUI);
		
		gallery	= createGallery(templates, commonData, listData);
	}

	/** build a Gallery from selected images */
	private Gallery createGallery(Templates templates, CommonData commonData, ImageListData listData) {
		final Gallery			 out	= new Gallery(templates, commonData);
		final List<ImageData>	todo	= listData.getSelected();
		for (Iterator<ImageData> it=todo.iterator(); it.hasNext();) {
			final ImageData	imageData	= it.next();
			final Upload		upload		= new Upload(commonData, imageData);
			out.addUpload(upload);
		}
		out.chain();
		return out;
	}

	@Override
	protected void execute() {
		if (gallery.isEmpty()) {
			log.info("nothing to upload");
			statusUILater.halt("status.upload.empty", new Object[] {});
			return; 
		}
		
		//### BÄH
		final String	wiki		= commonData.wiki;
		final String	user		= commonData.user;
		final String	password	= commonData.password;
		
		//------------------------------------------------------------------------------
		//## login
		
		log.info("logging in");
		try {
			// login
			statusUILater.indeterminate("status.login.started", new Object[] { wiki });
			final boolean	remember	= false;
			try {
				boolean success	= mw.login(wiki, user, password, remember);
				if (!success) { statusUILater.halt("status.login.wrongpw", new Object[] { wiki }); return; }
			}
			catch (Exception e) {
				statusUILater.halt("status.login.error", new Object[] { wiki, e.getMessage() });
				log.error("login failed", e);
				return;
			}
			
			check();
		}
		catch (AbortedException e) {
			log.info("login aborted");
			statusUILater.halt("status.login.aborted", new Object[0]);
			return;
		}
		statusUILater.halt("status.login.successful", new Object[] { wiki });
		
		//------------------------------------------------------------------------------
		//## uploads

		log.info("uploading files");
		try {
			// upload files and build gallery
			for (Iterator<Upload> it=gallery.getUploads().iterator(); it.hasNext();) {
				final Upload	upload	= it.next();
				final File	file	= upload.getFile();
				
				statusUILater.halt("status.upload.started", new Object[] { file.getPath() });
				final String	title	= upload.getTitle();
				final String	text	= gallery.imageDescription(upload);
				try {
					final ProgressCallback	progressCallback	= new MyProgressCallback(upload);
					final UploadCallback		uploadCallback		= new MyUploadCallback(upload);
					final boolean		watchThis	= true;	// TODO: should be configurable
					final Uploaded	uploaded	= mw.upload(wiki, title, text, file, watchThis, progressCallback, uploadCallback);
					final Location	uploc		= uploaded.location;
					// TODO uploc may be null if the user aborted
					statusUILater.halt("status.upload.successful", new Object[] { file.getPath(), uploc.toString() });
					upload.setLocation(uploc);
				}
				catch (Exception e) {
					statusUILater.halt("status.upload.error", new Object[] { file.getPath(), e.getMessage() });
					try {
						// HACK..
						final Location	uploc	= mw.absoluteLocation(
								wiki + ":" + mw.nameSpace(wiki, NameSpace.FILE).addTo(title));
						upload.setLocation(uploc);
						upload.setError(e);
						log.error("upload failed: " + title, e);
					}
					catch (UnsupportedWikiException e2) {
						log.error("gallery could not be changed", e2);
					}
				}
				
				check();
			}
			log.info("upload complete");
		}
		catch (TemplateException e) {
			log.error("template problem", e);
			statusUILater.halt("status.upload.aborted", new Object[0]);
			return;
		}
		catch (AbortedException e) {
			log.info("upload aborted");
			statusUILater.halt("status.upload.aborted", new Object[0]);
			return;
		}
		
		//------------------------------------------------------------------------------
		//## gallery

		log.info("changing gallery");
		try {
			// load gallery
			final Location	homeLocation	= mw.homeLocation(wiki);
			final Location	galleryLocation	= mw.relativeLocation(homeLocation, "/gallery");
			statusUILater.indeterminate("status.gallery.loading", new Object[] { "[[" + galleryLocation.toString() + "]]" });
			final Page		loaded			= mw.load(galleryLocation);
			
			// change gallery
			statusUILater.indeterminate("status.gallery.storing", new Object[] { "[[" + loaded.location.toString() + "]]" });
			final String	text	= gallery.galleryDescription();
			final Page		store	= loaded.edit(text + "\n\n" + TextUtil.trimLF(loaded.body));
			final String	summary	= gallery.gallerySummary();
			final boolean	minor	= false;
			final Page	stored	= mw.store(store, summary, minor);
			if (stored == null)	statusUILater.halt("status.gallery.updated",		new Object[] { "[[" + loaded.location.toString() + "]]" });
			else				statusUILater.halt("status.gallery.editConflict",	new Object[] { "[[" + loaded.location.toString() + "]]" });
		}
		catch (Exception e) {
			log.error("gallery update error");
			statusUILater.halt("status.gallery.error", new Object[] { e.getMessage() });
		}
		
//		//------------------------------------------------------------------------------
//		//## logout
//		
//		log.info("logging out");
//		try {
//			statusUILater.indeterminate("status.logout.started", new Object[] { wiki });
//			boolean success	= mw.logout(wiki);
//			if (!success) { statusUILater.halt("status.logout.failed", new Object[] { wiki }); return; }
//		}
//		catch (Exception e) {
//			statusUILater.halt("status.logout.error", new Object[] { wiki, e.getMessage() });
//			log.error("logout failed", e);
//			return;
//		}
//		statusUILater.halt("status.logout.successful", new Object[] { wiki });
		
		//------------------------------------------------------------------------------
		
		log.info("upload finished");
	}
	
	//------------------------------------------------------------------------------
	//## callbacks
	
	/** updates the statusUI when file upload progresses */
	private class MyProgressCallback implements ProgressCallback {
		private final Upload	upload;
		
		public MyProgressCallback(Upload upload) {
			this.upload	= upload;
		}

		public void bytesWritten(long bytes, long ofBytes) {
			// System.err.println("written " + bytes + " of " + ofBytes);
			final int	percent	= (int)(bytes * 100 / ofBytes);
			final String	path	= upload.getFile().getPath();
			statusUILater.determinate("status.upload.progress", new Object[] { path, new Integer(percent) }, percent, 100);
			// rate = (bytes - oldBytes) / (time  - oldTime)
		}
	}
	
	/** asks the user when somwethiung about a file upload is unclear */
	private class MyUploadCallback implements UploadCallback { 
		private final Upload	upload;
		
		public MyUploadCallback(Upload upload) {
			this.upload	= upload;
		}
		
		/** when true, the file overwites another file */
		public boolean ignoreFileexists() { 
			final String	name	= upload.getName();
			return callbackQuery(name, 
					"query.upload.ignoreFileexists.title", 
					"query.upload.ignoreFileexists.message");	// false
		}
		
		/** when non-null the result is used as new file name */
		public String renameFileexists() { 
			log.debug("renameFileexists"); 
			return null; 
		}
		
		/** if true, a large file will be written */
		public boolean ignoreLargefile() { 
			log.debug("ignoreLargefile"); 
			return true; 
		}
		
		/** if true, previously uploaded and deleted files will be uploaded nevertheless */
		public boolean ignoreFilewasdeleted() { 
			final String	name	= upload.getName();
			return callbackQuery(name, 
					"query.upload.ignoreFilewasdeleted.title", 
					"query.upload.ignoreFilewasdeleted.message");	// true
		}
		
		/** ask the user a yes/no message */
		private boolean callbackQuery(final String fileName, final String titleKey, final String messageKey) {
			final boolean[]	answer	= new boolean[] { false };
			try {
				SwingUtilities.invokeAndWait(new Runnable() {
					public void run() {
						answer[0]	= JOptionPane.YES_OPTION == JOptionPane.showConfirmDialog(
								mainWindow.window,
								Messages.message(messageKey, new Object[] { fileName }),
								Messages.text(titleKey),
								JOptionPane.YES_NO_OPTION);
					}
				});
			}
			catch (Exception e) {
				log.error("callback error", e);
				throw new RuntimeException(e);
			}
			return answer[0];
		}
	}
}
